alacritty_terminal/src/term/mod.rs

```diff
@@ -830,13 +920,122 @@ impl SizeInfo {
             col: min(col, Column(self.cols().saturating_sub(1))),
         }
     }
+
+    #[inline(always)]
+    pub fn into_u32(self) -> (u32, u32, u32, u32, u32, u32) {
+        (
+            self.width as u32,
+            self.height as u32,
+            self.cell_width as u32,
+            self.cell_height as u32,
+            self.padding_x as u32,
+            self.padding_y as u32,
+        )
+    }
 }
 
 impl<T> Term<T> {
     pub fn selection(&self) -> &Option<Selection> {
         &self.grid.selection
     }
 
+    fn calculate_selection_damage(
+        &self,
+        cols: index::Column,
+        selection: &Option<Selection>,
+    ) -> Option<(Column, Line, Column, Line)> {
+        match selection {
+            Some(ref selection) => match selection.to_span(self) {
+                Some(span) => {
+                    let (limit_start, limit_end) = if span.is_block {
+                        (span.end.col, span.start.col)
+                    } else {
+                        (Column(0), self.cols() - 1)
+                    };
+
+                    // Get on-screen lines of the selection's locations
+                    let mut start = self.buffer_to_visible(span.start);
+                    let mut end = self.buffer_to_visible(span.end);
+
+                    // Start and end lines are swapped as we switch from buffer to line coordinates
+                    if start > end {
+                        mem::swap(&mut start, &mut end);
+                    }
+
+                    let (start_col, end_col) = if start.line == end.line {
+                        // Trim start/end with partially visible block selection
+                        (max(limit_start, start.col), min(limit_end, end.col))
+                    } else {
+                        // Multiple lines, so do full width damage
+                        (Column(0), cols - 1)
+                    };
+
+                    Some((start_col, Line(start.line), end_col, Line(end.line)))
+                }
+                None => None,
+            }
+            None => None,
+        }
+    }
+
+    /// Finalize and retrieve current damage. Note that the user is responsible
+    /// for calling LineDamage::reset on every line returned for efficiency.
+    pub fn get_damage(&mut self) -> (bool, &mut Vec<LineDamage>) {
```

Comment:
I don't like the return of this function. Could we somehow only return mut ref to vector?

Reply:
No, but we could maybe return a mut ref to the damage struct.

Otherwise, if full damage is set, the vector is invalid. We'd have to iterate over it and write full damage to every line if we wanted to avoid returning the full damage parameter.

Reply:
Lets just return a mut ref to damage struct, or store full damage on Term. I don't like when get_damage returns <bool, Damage>. I don't have any better idea for now.